// fluids/dist/cjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeFluidObserver = exports.addFluidObserver = exports.setFluidGetter = exports.callFluidObservers = exports.callFluidObserver = exports.getFluidObservers = exports.getFluidValue = exports.hasFluidValue = exports.FluidValue = void 0;
var $get = Symbol.for('FluidValue.get');
var $observers = Symbol.for('FluidValue.observers');
/** Returns true if `arg` can be observed. */
var hasFluidValue = function (arg) { return Boolean(arg && arg[$get]); };
exports.hasFluidValue = hasFluidValue;
/**
 * Get the current value.
 * If `arg` is not observable, `arg` is returned.
 */
var getFluidValue = function (arg) {
    return arg && arg[$get] ? arg[$get]() : arg;
};
exports.getFluidValue = getFluidValue;
/** Get the current observer set. Never mutate it directly! */
var getFluidObservers = function (target) {
    return target[$observers] || null;
};
exports.getFluidObservers = getFluidObservers;
function callFluidObserver(observer, event) {
    if (observer.eventObserved) {
        observer.eventObserved(event);
    }
    else {
        observer(event);
    }
}
exports.callFluidObserver = callFluidObserver;
function callFluidObservers(target, event) {
    var observers = target[$observers];
    if (observers) {
        observers.forEach(function (observer) {
            callFluidObserver(observer, event);
        });
    }
}
exports.callFluidObservers = callFluidObservers;
/**
 * Extend this class for automatic TypeScript support when passing this
 * value to `fluids`-compatible libraries.
 */
var FluidValue = /** @class */ (function () {
    function FluidValue(get) {
        if (!get && !(get = this.get)) {
            throw Error('Unknown getter');
        }
        setFluidGetter(this, get);
    }
    return FluidValue;
}());
exports.FluidValue = FluidValue;
/** Define the getter called by `getFluidValue`. */
var setFluidGetter = function (target, get) {
    return setHidden(target, $get, get);
};
exports.setFluidGetter = setFluidGetter;
function addFluidObserver(target, observer) {
    if (target[$get]) {
        var observers = target[$observers];
        if (!observers) {
            setHidden(target, $observers, (observers = new Set()));
        }
        if (!observers.has(observer)) {
            observers.add(observer);
            if (target.observerAdded) {
                target.observerAdded(observers.size, observer);
            }
        }
    }
    return observer;
}
exports.addFluidObserver = addFluidObserver;
function removeFluidObserver(target, observer) {
    var observers = target[$observers];
    if (observers && observers.has(observer)) {
        var count = observers.size - 1;
        if (count) {
            observers.delete(observer);
        }
        else {
            setHidden(target, $observers, null);
        }
        if (target.observerRemoved) {
            target.observerRemoved(count, observer);
        }
    }
}
exports.removeFluidObserver = removeFluidObserver;
var setHidden = function (target, key, value) {
    return Object.defineProperty(target, key, {
        value: value,
        writable: true,
        configurable: true,
    });
};
